### PR TITLE
Returned cached Brave Ads preference value if sync mojom call failed.

### PR DESCRIPTION
**Pull request by aseren**
The current PR also does the following:

* Returns cached Brave Ads preference value if sync mojom call failed.
* Removes Bat Ads Service restart on mojom disconnection. This is not needed because we are running Bat Ads in the browser process.

Resolves https://github.com/brave/brave-browser/issues/34391

## Submitter Checklist:

- [x] I confirm that no [security/privacy review is needed](https://github.com/brave/brave-browser/wiki/Security-reviews) and no other type of reviews are needed, or that I have [requested](https://github.com/brave/reviews/issues/new/choose) them
- [x] There is a [ticket](https://github.com/brave/brave-browser/issues) for my issue
- [x] Used Github [auto-closing keywords](https://docs.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue) in the PR description above
- [x] Wrote a good [PR/commit description](https://google.github.io/eng-practices/review/developer/cl-descriptions.html)
- [x] Squashed any review feedback or "fixup" commits before merge, so that history is a record of what happened in the repo, not your PR
- [x] Added appropriate labels (`QA/Yes` or `QA/No`; `release-notes/include` or `release-notes/exclude`; `OS/...`) to the associated issue
- [x] Checked the PR locally:
  * `npm run test -- brave_browser_tests`, `npm run test -- brave_unit_tests` [wiki](https://github.com/brave/brave-browser/wiki/Tests)
  * `npm run lint`, `npm run presubmit` [wiki](https://github.com/brave/brave-browser/wiki/Presubmit-checks), `npm run gn_check`, `npm run tslint`
- [ ] Ran `git rebase master` (if needed)

## Reviewer Checklist:

- [ ] A security review [is not needed](https://github.com/brave/brave-browser/wiki/Security-reviews), or a link to one is included in the PR description
- [ ] New files have MPL-2.0 license header
- [ ] Adequate test coverage exists to prevent regressions
- [ ] Major classes, functions and non-trivial code blocks are well-commented
- [ ] Changes in component dependencies are properly reflected in `gn`
- [ ] Code follows the [style guide](https://chromium.googlesource.com/chromium/src/+/HEAD/styleguide/c++/c++.md)
- [ ] Test plan is specified in PR before merging

## After-merge Checklist:

- [ ] The associated issue milestone is set to the smallest version that the
  changes has landed on
- [ ] All relevant documentation has been updated, for instance:
  - [ ] https://github.com/brave/brave-browser/wiki/Deviations-from-Chromium-(features-we-disable-or-remove)
  - [ ] https://github.com/brave/brave-browser/wiki/Proxy-redirected-URLs
  - [ ] https://github.com/brave/brave-browser/wiki/Fingerprinting-Protections
  - [ ] https://github.com/brave/brave-browser/wiki/Brave%E2%80%99s-Use-of-Referral-Codes
  - [ ] https://github.com/brave/brave-browser/wiki/Custom-Headers
  - [ ] https://github.com/brave/brave-browser/wiki/Web-Compatibility-Exceptions-in-Brave
  - [ ] https://github.com/brave/brave-browser/wiki/QA-Guide
  - [ ] https://github.com/brave/brave-browser/wiki/P3A

## Test Plan:

The PR fixes a rare intermittent crash which is hard to reproduce.

### CODE REVIEW

File: components/services/bat_ads/bat_ads_client_mojo_bridge.h [hunk 1]
```diff
@@ -119,6 +119,13 @@ class BatAdsClientMojoBridge : public brave_ads::AdsClient {
            const std::string& message) override;
 
  private:
+  absl::optional<base::Value> GetCachedProfilePrefValue(
```
Comment (tmancey) [inline components/services/bat_ads/bat_ads_client_mojo_bridge.h:122]: nit: `CachedProfilePrefValue`

File: components/services/bat_ads/bat_ads_client_mojo_bridge.h [hunk 2]
```diff
@@ -119,6 +119,13 @@ class BatAdsClientMojoBridge : public brave_ads::AdsClient {
...
+      const std::string& path) const;
+  absl::optional<base::Value> GetCachedLocalStatePrefValue(
```
Comment (tmancey) [inline components/services/bat_ads/bat_ads_client_mojo_bridge.h:124]: nit: `CachedLocalStatePrefValue`

File: components/services/bat_ads/bat_ads_client_mojo_bridge.h [hunk 3]
```diff
@@ -119,6 +119,13 @@ class BatAdsClientMojoBridge : public brave_ads::AdsClient {
...
+      const std::string& path) const;
+
+  base::flat_map<std::string, base::Value> cached_profile_prefs_;
+  base::flat_map<std::string, base::Value> cached_local_state_prefs_;
```
Comment (tmancey) [inline components/services/bat_ads/bat_ads_client_mojo_bridge.h:128]: nit: comments for `std::string` and `base::Value`, i.e., `/*path=*/` and `/*value=*/`

File: components/services/bat_ads/bat_ads_client_mojo_bridge.cc [hunk 4]
```diff
@@ -290,59 +290,84 @@ void BatAdsClientMojoBridge::Log(const char* file,
...
+    return GetCachedProfilePrefValue(path);
   }
 
   absl::optional<base::Value> value;
-  bat_ads_client_associated_receiver_->GetProfilePref(path, &value);
+  if (!bat_ads_client_associated_receiver_->GetProfilePref(path, &value)) {
+    return GetCachedProfilePrefValue(path);
+  }
+
+  if (value) {
```
Comment (tmancey) [inline components/services/bat_ads/bat_ads_client_mojo_bridge.cc:303]: would the above condition not fail if value is null?

File: components/services/bat_ads/bat_ads_client_mojo_bridge.cc [hunk 5]
```diff
@@ -290,59 +290,84 @@ void BatAdsClientMojoBridge::Log(const char* file,
...
+    return GetCachedLocalStatePrefValue(path);
   }
 
   absl::optional<base::Value> value;
-  bat_ads_client_associated_receiver_->GetLocalStatePref(path, &value);
+  if (!bat_ads_client_associated_receiver_->GetLocalStatePref(path, &value)) {
+    return GetCachedLocalStatePrefValue(path);
+  }
+
+  if (value) {
```
Comment (tmancey) [inline components/services/bat_ads/bat_ads_client_mojo_bridge.cc:351]: See comment for `GetProfilePref`

File: components/services/bat_ads/bat_ads_client_mojo_bridge.cc [hunk 6]
```diff
@@ -351,12 +376,35 @@ void BatAdsClientMojoBridge::ClearLocalStatePref(const std::string& path) {
...
+  }
+
   return value;
 }
 
+absl::optional<base::Value> BatAdsClientMojoBridge::GetCachedProfilePrefValue(
+    const std::string& path) const {
+  if (!cached_profile_prefs_.contains(path)) {
+    return absl::nullopt;
+  }
```
Comment (tmancey) [inline components/services/bat_ads/bat_ads_client_mojo_bridge.cc:395]: Could we not call `HasProfilePref`?

File: components/services/bat_ads/bat_ads_client_mojo_bridge.cc [hunk 7]
```diff
@@ -351,12 +376,35 @@ void BatAdsClientMojoBridge::ClearLocalStatePref(const std::string& path) {
...
+
+  return cached_profile_prefs_.at(path).Clone();
+}
+
+absl::optional<base::Value>
+BatAdsClientMojoBridge::GetCachedLocalStatePrefValue(
+    const std::string& path) const {
+  if (!cached_local_state_prefs_.contains(path)) {
+    return absl::nullopt;
+  }
```
Comment (tmancey) [inline components/services/bat_ads/bat_ads_client_mojo_bridge.cc:405]: Could we not call `HasLocalStatePref`?

**Review by tmancey — APPROVED** [review]
LGTM++ (Please see comments)

File: components/services/bat_ads/bat_ads_client_mojo_bridge.cc (hunk 4 above)
Comment (aseren) [reply components/services/bat_ads/bat_ads_client_mojo_bridge.cc:303]: yes, it is possible based on our Mojom interface definition:

```
GetProfilePref(string path) => (mojo_base.mojom.Value? value);
```

File: components/services/bat_ads/bat_ads_client_mojo_bridge.cc (hunk 5 above)
Comment (aseren) [reply components/services/bat_ads/bat_ads_client_mojo_bridge.cc:351]: nullopt here is also possible based on our Mojom interface definition:

```
GetLocalStatePref(string path) => (mojo_base.mojom.Value? value);
```

**Review by kylehickinson — APPROVED** [review]
iOS++

**Pull request merged**